### ISSUE TITLE
Add event.composedPath()

### ISSUE THREAD

**Issue by amsul** [jsdom/jsdom#1563]
Events dispatched through jsdom seem to be missing the `path` property: https://dom.spec.whatwg.org/#event-path

> An event has an associated path. A path is a list of tuples, each of which consists of an item (an EventTarget object) and a target (null or an EventTarget object). A tuple is formatted as (item, target). A path is initially the empty list.

Not quite sure what it would require to support this, but if you could give me some pointers, I can attempt an implementation.

Is there any other API jsdom provides that would help determine the path an event is bubbling through?

**Comment by domenic** [comment]
There is no event.path property. The part of the spec you quote is just defining the concept of an event path.

There is a composedPath() method: https://dom.spec.whatwg.org/#dom-event-composedpath

However, this is not implemented in any browsers yet, so I'm not sure jsdom should go first there. But a pull request would be accepted if it had enough tests.

> Is there any other API jsdom provides that would help determine the path an event is bubbling through?

How would you figure this out in a browser today?

**Comment by amsul** [comment]
Chrome has the `event.path` property defined.. I naively assumed it was widely supported. But you are correct about the `composedPath()` method (although that doesn't seem to be defined in Chrome).

And yes, I agree, if it's not widely supported yet, jsdom doesn't need to be chaperoning it forward.

Honestly, I am not sure what the "proper" way to do this is in a browser today. The most basic approach would probably be to just climb up the `.parentNode` references until you reach the event target.

**Comment by amsul** [comment]
Okay, for anyone wondering, this seems to be a decent enough workaround for now:

```js
function getEventPath(event) {

  if (event.path) {
    return event.path
  }

  let path = []
  let target = event.target

  while (target.parentNode) {
    path.push(target)
    target = target.parentNode
  }

  path.push(document, window)

  return path

}
```

@domenic feel free to close this as it is beyond the scope of jsdom atm.

**Comment by tomalec** [comment]
Just to update, it seems latest Chrome supports `composedPath` http://jsbin.com/vepohu/3/edit?html,console,output

**Comment by Zirro** [comment]
This was implemented as part of https://github.com/jsdom/jsdom/pull/2347, and got released in v12.2.0. 🎉
